packages/logging/src/metadata.js

```diff
@@ -0,0 +1,297 @@
+ * @returns {Object}
+ */
+Metadata.prototype.produceGAEDescriptor_ = function(projectId) {
+  return {
+    type: 'gae_app',
+    labels: {
+      project_id: projectId,
+      module_id: process.env.GAE_SERVICE,
+      version_id: process.env.GAE_VERSION
+    }
+  };
+};
+
+/**
+ * Called to create a descriptor for Google Cloud Functions
+ *
+ * @private
+ * @param {String} projectId - the project id
+ * @returns {Object}
+ */
+Metadata.prototype.produceCFDescriptor_ = function(projectId) {
+  return {
+    type: 'cloud_function',
+    labels: {
+      project_id: projectId,
+      function_name: process.env.FUNCTION_NAME,
+      region: process.env.SUPERVISOR_REGION
+    }
+  };
+};
+
+/**
+ * Called to create a global descriptor
+ *
+ * @private
+ * @param {String} projectId - the project id
+ * @returns {Object}
+ */
+Metadata.prototype.produceGlobalDescriptor_ = function(projectId) {
+  return {
+    type: 'global',
+    labels: {
+      project_id: projectId
+    }
+  };
+};
+
+/**
+ * Invoke to attempt to retrieve a resource object describing the current
+ * environment.
+ *
+ * @param {Function} callback - the error-back style callback to invoke once
+ *  the resource has been determined.
+ * @returns {module:metadata}
+ * @chainable
+ */
+Metadata.prototype.getDefaultResource = function(callback) {
+  var self = this;
+  if (!is.function(callback)) {
+    throw new TypeError(
+      'Must supply a callback function as an argument to invocation.');
+  }
+  this.getProjectId(function(err, projectId) {
+    if (err) {
+      callback(err, null);
+      return;
+    }
+    if (!is.null(self.metadataProjectId_)) {
+      switch (true) {
+        case self.isCloudFunction_(): {
+          callback(null, self.produceCFDescriptor_(projectId));
+          break;
+        }
+        case self.isAppEngine_(): {
+          callback(null, self.produceGAEDescriptor_(projectId));
+          break;
+        }
+        default: {
+          // Since the metadata service is reachable assume it's GCE
+          callback(null, self.produceGCEDescriptor_(projectId));
+        }
+      }
+      return;
+    }
+    // Produce a global descriptor due to the lack of information available
+    callback(null, self.produceGlobalDescriptor_(projectId));
+  });
+  return this;
+};
+
+/**
+ * Invoke to attempt to retrieve the project id from the current environment or
+ * from the metadata service.
+ *
+ * @param {Function} callback - the error-back style callback to invoke once
+ *  the project id has been determined.
+ * @returns {module:metadata}
+ * @chainable
+ */
+Metadata.prototype.getProjectId = function(callback) {
+  var self = this;
+  if (!is.function(callback)) {
+    throw new TypeError(
+      'Must supply a callback function as an argument to invocation.');
+  } else if (!is.empty(self.processProjectId_()) &&
+    is.string(self.processProjectId_())) {
+    // The process's GCLOUD_PROJECT env variable has been set
```

Comment:
Sure, but in that case the default resource would still get written to the entry. In essence this wouldn't affect a user-supplied project id in the resource field, since: if one is not provided it will be automatically provided and if one is provided by the user then that one will overwrite the auto generated one in the resource